docs/source/daml2ts/index.rst

```diff
@@ -0,0 +1,290 @@
+| ``Date``          | ``Date``         | ``string``                       |
++-------------------+------------------+----------------------------------+
+| ``ContractId<T>`` | ``ContractId t`` | ``string``                       |
++-------------------+------------------+----------------------------------+
+| ``Optional<T>``   | ``Optional t``   | ``null | (null extends T ?``     |
+|                   |                  | ``[] | [Exclude<T, null>] : T)`` |
++-------------------+------------------+----------------------------------+
+| ``TextMap<T>``    | ``TextMap t``    | ``{ [key: string]: T }``         |
++-------------------+------------------+----------------------------------+
+
+.. note::
+   The TypeScript types ``Time``, ``Decimal``, ``Numeric`` and ``Int`` all alias to ``string``. These choices relate to the avoidance of precision loss under serialization over the `json-api <../json-api/index.html>`_.
+
+.. note::
+   The Typescript definition of type ``Optional<T>`` in the above table might look complicated. It accounts for differences in the encoding of optional values when nested versus when they are not (i.e. "top-level"). For example, ``null`` and ``"foo"`` are two possible values of ``Optional<Text>`` whereas, ``[]`` and ``["foo"]`` are two possible values of type ``Optional<Optional<Text>>`` (``null`` is another possible value, ``[null]`` is **not**).
+
+DAML to TypeScript mappings
+---------------------------
+
+The mappings from DAML to TypeScript are best explained by example.
+
+Records
+~~~~~~~
+
+In DAML, we might model a person like this.
+
+.. code-block:: daml
+   :linenos:
+
+   data Person =
+     Person with
+       name: Text
+       party: Party
+       age: Int
+
+Given the above definition, the generated TypeScript code will be as follows.
+
+.. code-block:: typescript
+   :linenos:
+
+   type Person = {
+     name: string;
+     party: daml.Party;
+     age: daml.Int;
+   }
+
+Tuples
+~~~~~~
+
+Suppose the following DAML definition.
+
+.. code-block:: daml
+   :linenos:
+
+   data Pair a b = Pair (a, b)
+
+In TypeScript, the above definition renders as below [1]_.
+
+.. code-block:: typescript
+   :linenos:
+
+   type Pair<a, b> =
+     |  { tag: 'Pair'; value: {_1: a, _2: b} }
+
+
+.. [1] This is a simplification but it is not wrong to think of it this way.
+
+.. note::
+   The general rule is that DAML tuples are represented in TypeScript as simple objects with fields labeled ``_1``, ``_2`` and so on.
+
+The DAML type ``Pair a b`` has been modeled as a TypeScript discriminated union (with a single case). In general, all DAML variants are translated so as we will see in the next section.
+
+Variants
+~~~~~~~~
+
+This is a DAML type for a language of additive expressions.
+
+.. code-block:: daml
+   :linenos:
+
+   data Expr a =
+       Lit a
+     | Var Text
+     | Add (Expr a, Expr a)
+
+In TypeScript, it is represented as a discriminated union.
```

Comment:
It might make sense to put a link to the "Discriminated Unions" section in the TS handbook under the phrase "discriminated union".
https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions

Reply:
✔️